ADR 016: Validator consensus key rotation (#5233)
cosmos/cosmos-sdk: docs/architecture/adr-016-validator-consensus-key-rotation.md

# ADR 016: Validator Consensus Key Rotation

## Changelog

- 2019 Oct 23: Initial draft

## Context

Validator consensus key rotation feature has been discussed and requested for a long time, for the sake of safer validator 
key management policy (e.g. https://github.com/tendermint/tendermint/issues/1136). So, we suggest one of the simplest form of
validator consensus key rotation implementation mostly onto Cosmos-SDK.

## Decision

### Pseudo procedure for consensus key rotation

- create new random consensus key.
- create and broadcast a transaction with a `MsgRotateConsPubKey` that states the new consensus key is now coupled with the validator operator with signature from the validator's operator key.
- old consensus key becomes unable to participate on consensus immediately after the update of key mapping state on-chain.
- start validating with new consensus key.

> - validators using HSM and KMS should update the consensus key in HSM to use the new rotated key after the height `h` when `MsgRotateConsPubKey` committed to the blockchain.

Reviewer: I opened a KMS issue about supporting more than one registered key per chain, and ways that Tendermint could signal which key to use (e.g. if it detects a misconfiguration or invalid signature when requesting a signature from the new key):

https://github.com/tendermint/kms/issues/368

I think it's important to be able to handle failures during key rotations without requiring manual intervention, although perhaps it isn't required for a key rotation MVP. Without a single place to coordinate flipping which key is active, the rollback procedure will be much more complex, IMO.

Author: 1)

If i am understanding right, tendermint/kms#368 seems similar to something like an OTP, even though it will be not practically one time.

Trigger of automatic key rotation can be decided by latest block hash. We can encode such block hash into an integer, and then key rotation is decided by the remainder of the integer divided by some high integer number. Also we can design an algorithm so that the rotation happens statistically every N hours or N days.

But these ideas are too complex to put it in this ADR at the moment.

2) There will be definitely an only struct in SDK which stores the ChangeLog of consensus key rotation. But a rollback on this data seems very risky, because there exists high possibility that the validator has either lost or been stolen the key. So, intuitively, even though the state of chain being rolled back, the consensus key for rolled back chain on genesis should not be rolled back but should be provided by each validator manually with their signature by the operator key.

I cannot think of any simple automatic solution to this yet.

Reviewer 2: That seems overly complex and it's not immediately clear to me how that would work. As far as I see it, one of the main issues to address is once the Tx is committed in a block, KMS will somehow have to be notified automatically (or manually?) that it should use the new key in the keyring.

Author: I think notifying KMS is strictly KMS related job on tendermint, and it is quite independent from this initial key rotation feature on "Cosmos-SDK". Our intention is to deploy the simplest form of rotation feature on "Cosmos-SDK" so that we make no change on tendermint.

I think, after implementation of this feature, KMS expert (like @tarcieri) can hack Cosmos-SDK to support such notification by monitoring the mapping struc between valoper and valconspub. It should monitor a struc in SDK because the mapping information is not stored in tendermint.

Reviewer: Ideally I think Tendermint would include the public key to use for the signature in the `SignProposal` and `SignVote` messages, although it seems that may not work for an MVP.

Author: Yes. This ADR includes only the simplest form of consensus key rotation without considering multiple consensus key situation. We think it should be remain as a long term goal to pursue because it needs a breaking structural change on Tendermint about validator and consensus key abstract definition.

Author: adding below statement in the context section

_Also, it should be noted that this ADR includes only the simplest form of consensus key rotation without considering multiple consensus keys concept. Such multple consensus keys concept shall be remained as a long term goal of Tendermint and Cosmos-SDK._

> - validators using HSM and KMS should update the consensus key in HSM to use the new rotated key after the height `h` when `MsgRotateConsPubKey` committed to the blockchain.

Reviewer 3: where height `h` comes from?

Author: height `h` = when `MsgRotateConsPubKey` committed to the blockchain.

### Considerations

- consensus key mapping information management strategy
    - store history of each key mapping changes in the kvstore.
    - the state machine can search corresponding consensus key paired with given validator operator for any arbitrary height in a recent unbonding period.
    - the state machine does not need any historical mapping information which is past more than unbonding period.
- limits
    - a validator cannot rotate its consensus key more than N time for any unbonding period, to prevent spam.
    - parameters can be decided by governance and stored in genesis file.
- slash module
    - slash module can search corresponding consensus key for any height so that it can decide which consensus key is supposed to be used for given height.
- abci.ValidatorUpdate
    - tendermint already has ability to change a consensus key by ABCI communication(`ValidatorUpdate`).
    - validator consensus key update can be done via creating new + delete old by change the power to zero.
    - therefore, we expect we even do not need to change tendermint codebase at all to implement this feature.

> 

Reviewer 3: what about charging validators for rotating a key as per @cwgoes comment?

```
However, frequent key rotations will make bisection less efficient, since fewer headers will be able to be skipped by the light client, so limiting the rate of key change is prudent. I think it might be sufficient to charge validators some nominal amount for changing their key, which could be directed to the community pool - that would "internalize the externality" of more expensive light client verification. Rate limiting to once per unbonding period or comparable duration is also probably fine.
```
https://github.com/tendermint/tendermint/issues/1136#issuecomment-542656685

Author: It was there in original suggestion but it was removed.
I am quite neutral to the charging.
But I think if we charge cost, I think the fee should be higher when number of rotation for given period is high.

- suggestion example : fee = 2 ^ (# of rotation for recent unbonding period) atoms
so, for 10th time in recent unbonding period, it costs 2^10 = 1024 atoms

Reviewer 3: Agree re exponential cost.

Reviewer 2: Yes, the idea is to have an on-chain parameter limiting the total (N) number of rotations per unbonding period. In addition, we can specifically charge higher gas costs for this specific Tx.

Author: we will update the spec as below
fee = initial_fee * 2^n 
(n = # of rotation for recent unbonding period, not including this one)
the initial_fee can be decided by governance and our suggestion is 1 atom.
Does it sound OK?

Author: Updated.

    ```go
    type ConsPubKeyRotationHistory struct {
        OperatorAddress         sdk.ValAddress
        OldConsPubKey           crypto.PubKey
        NewConsPubKey           crypto.PubKey
        RotatedHeight           int64
    }
    ```

5. `ApplyAndReturnValidatorSetUpdates` checks if there is `ConsPubKeyRotationHistory` with `ConsPubKeyRotationHistory.RotatedHeight == ctx.BlockHeight()` and if so, generates 2 `ValidatorUpdate` , one for a remove validator and one for create new validator 

    ```go
    abci.ValidatorUpdate{
        PubKey: tmtypes.TM2PB.PubKey(OldConsPubKey),
        Power:  0,
    }

    abci.ValidatorUpdate{
        PubKey: tmtypes.TM2PB.PubKey(NewConsPubKey),
        Power:  v.ConsensusPower(),
    }
    ```

6. at `previousVotes` Iteration logic of `AllocateTokens`,  `previousVote` using `OldConsPubKey` match up with `ConsPubKeyRotationHistory`, and replace validator for token allocation
7. Migrate `ValidatorSigningInfo` and `ValidatorMissedBlockBitArray` from `OldConsPubKey` to `NewConsPubKey`

## Status

Proposed

## Consequences

### Positive

- Validators can immediately or periodically rotate their consensus key to have better security policy

> 

Reviewer 3: ```suggestion
- improved security against Long-Range attacks (https://nearprotocol.com/blog/long-range-attacks-and-a-new-fork-choice-rule) given a validator throws away the old consensus key(s)
```

### Negative

- Slash module needs more computation because it needs to lookup corresponding consensus key of validators for each height

> 

Reviewer 3: ```suggestion
- frequent key rotations will make light client bisection less efficient
```

# ADR 016: Validator Consensus Key Rotation

## Changelog

- 2019 Oct 23: Initial draft
- 2019 Nov 28: Add key rotation fee

## Context

Validator consensus key rotation feature has been discussed and requested for a long time, for the sake of safer validator 
key management policy (e.g. https://github.com/tendermint/tendermint/issues/1136). So, we suggest one of the simplest form of

> validator consensus key rotation implementation mostly onto Cosmos-SDK.

Author: thanks for noting that. I add below sentences to clarify the reason.

_ We don't need to make any update on consensus logic in Tendermint because Tendermint does not have any mapping information of consensus key and validator operator key, meaning that from Tendermint point of view, a consensus key rotation of a validator is simply a replacement of a consensus key to another._

- start validating with new consensus key.
- validators using HSM and KMS should update the consensus key in HSM to use the new rotated key after the height `h` when `MsgRotateConsPubKey` committed to the blockchain.

### Considerations

- consensus key mapping information management strategy
    - store history of each key mapping changes in the kvstore.
    - the state machine can search corresponding consensus key paired with given validator operator for any arbitrary height in a recent unbonding period.
    - the state machine does not need any historical mapping information which is past more than unbonding period.
- limits
    - a validator cannot rotate its consensus key more than `MaxConsPubKeyRotations` time for any unbonding period, to prevent spam.
    - parameters can be decided by governance and stored in genesis file.
- key rotation fee
    - a validator should pay `KeyRotationFee` to rotate the consensus key which is calculated as below
    - `KeyRotationFee` = `InitialKeyRotationFee` * 2^(number of rotations in `ConsPubKeyRotationHistory` in recent unbonding period)
- slash module

>     - slash module can search corresponding consensus key for any height so that it can decide which consensus key is supposed to be used for given height.

Reviewer 4: the double sign evidence is now handled by the `x/evidence` module, which gets the registered pubkey from the `x/slashing` module. So ultimately the evidence module will have to look up the corresponding consensus address at the given height. 

cc: @alexanderbez

Author: thanks for pointing that out! as you said, evidence module gets pubkey from slashing keeper to use for double signing detection. edited as below

_- evidence module
    - evidence module can search corresponding consensus key for any height from slashing keeper so that it can decide which consensus key is supposed to be used for given height._

- abci.ValidatorUpdate
    - tendermint already has ability to change a consensus key by ABCI communication(`ValidatorUpdate`).
    - validator consensus key update can be done via creating new + delete old by change the power to zero.
    - therefore, we expect we even do not need to change tendermint codebase at all to implement this feature.

> - new genesis parameters

Reviewer 4: what are the default values proposed for these parameters?

Author: MaxConsPubKeyRotations : 10(11th try will be rejected)
InitialKeyRotationFee : 1atom(for the first rotation)

Reviewer 4: Can you add that to the ADR?

Author: Sure!

Author: done!

    - `MaxConsPubKeyRotations` : maximum number of rotation can be executed by a validator in recent unbonding period.
    - `InitialKeyRotationFee` : the initial key rotation fee when no key rotation has happened in recent unbonding period.

> ### Workflow

Reviewer 4: I'm confused about which module will this feature be a part of. I assume staking? Can you add that as a comment?

Author: I added two parts as below

1) in Considerations
- new genesis parameters _in `staking` module_

2) in Workflow
_- Note : All above features shall be implemented in `staking` module._

- evidence module
    - evidence module can search corresponding consensus key for any height from slashing keeper so that it can decide which consensus key is supposed to be used for given height.
- abci.ValidatorUpdate
    - tendermint already has ability to change a consensus key by ABCI communication(`ValidatorUpdate`).
    - validator consensus key update can be done via creating new + delete old by change the power to zero.
    - therefore, we expect we even do not need to change tendermint codebase at all to implement this feature.
- new genesis parameters in `staking` module
    - `MaxConsPubKeyRotations` : maximum number of rotation can be executed by a validator in recent unbonding period. default value 10 is suggested(11th key rotation will be rejected)
    - `InitialKeyRotationFee` : the initial key rotation fee when no key rotation has happened in recent unbonding period. default value 1atom is suggested(1atom fee for the first key rotation in recent unbonding period)

### Workflow

1. The validator generates a new consensus keypair.
2. The validator generates and signs a `MsgRotateConsPubKey` tx with their operator key and new ConsPubKey

    ```go

>     type MsgRotateConsPubKey struct {

Reviewer 5: Who is the signer of this message?

Author: The validator account key is the signer of `MsgRotateConsPubKey`

        ValidatorAddress  sdk.ValAddress
        NewPubKey         crypto.PubKey
    }
    ```

3. `handleMsgRotateConsPubKey` gets `MsgRotateConsPubKey`, calls `RotateConsPubKey` with emits event
4. `RotateConsPubKey` 
    - checks if `NewPubKey` is not duplicated on `ValidatorsByConsAddr`
    - checks if the validator is does not exceed parameter `MaxConsPubKeyRotations` by iterating `ConsPubKeyRotationHistory`
    - checks if the signing account has enough balance to pay `KeyRotationFee`
    - pays `KeyRotationFee` to community fund
    - overwrites `NewPubKey` in `validator.ConsPubKey`
    - deletes old `ValidatorByConsAddr`
    - `SetValidatorByConsAddr` for `NewPubKey`
    - Add `ConsPubKeyRotationHistory` for tracking rotation

    ```go
    type ConsPubKeyRotationHistory struct {
        OperatorAddress         sdk.ValAddress
        OldConsPubKey           crypto.PubKey
        NewConsPubKey           crypto.PubKey
        RotatedHeight           int64
    }
    ```

> 5. `ApplyAndReturnValidatorSetUpdates` checks if there is `ConsPubKeyRotationHistory` with `ConsPubKeyRotationHistory.RotatedHeight == ctx.BlockHeight()` and if so, generates 2 `ValidatorUpdate` , one for a remove validator and one for create new validator 

Reviewer 5: Will this change the priority of the validator in the RoundRobin proposer selection in Tendermint?

Author: The new rotated key will have the lowest accum because it is seen as a new validator. So it is a slight disadvantage to the key rotated validator

    ```go
    abci.ValidatorUpdate{
        PubKey: tmtypes.TM2PB.PubKey(OldConsPubKey),
        Power:  0,
    }

    abci.ValidatorUpdate{
        PubKey: tmtypes.TM2PB.PubKey(NewConsPubKey),
        Power:  v.ConsensusPower(),
    }
    ```

6. at `previousVotes` Iteration logic of `AllocateTokens`,  `previousVote` using `OldConsPubKey` match up with `ConsPubKeyRotationHistory`, and replace validator for token allocation
7. Migrate `ValidatorSigningInfo` and `ValidatorMissedBlockBitArray` from `OldConsPubKey` to `NewConsPubKey`
- Note : All above features shall be implemented in `staking` module.

## Status

Proposed

## Consequences

### Positive

- Validators can immediately or periodically rotate their consensus key to have better security policy
- improved security against Long-Range attacks (https://nearprotocol.com/blog/long-range-attacks-and-a-new-fork-choice-rule) given a validator throws away the old consensus key(s)

> ### Negative

Reviewer 6: Missing an important negative of it allows a validator to effectively sell their entity, in a way that before would require hardware assumptions. (They can change their key to give it to new management, without any oversight from their delegators)

# ADR 016: Validator Consensus Key Rotation

## Changelog

- 23-10-2019: initial draft

## Context

Validator consensus key rotation feature has been discussed and requested for a long time, for the sake of safer validator 
key management policy (e.g. https://github.com/tendermint/tendermint/issues/1136). So, we suggest one of the simplest form of
validator consensus key rotation implementation mostly onto Cosmos-SDK.

## Decision

### Pseudo procedure for consensus key rotation

- create new random consensus key.

> - create and broadcast a transaction(RotateValConsensusKey) that the new consensus key is now coupled with the validator operator with signature from validator wallet key.

Reviewer 2: ```suggestion
- create and broadcast a transaction with a `MsgRotateConsPubKey` that states the new consensus key is now coupled with the validator operator with signature from the validator's operator key.
```

- old consensus key becomes unable to participate on consensus immediately after the update of key mapping state on-chain.
- start validating with new consensus key.

> - validators using HSM and KMS should update the consensus key in HSM to use the new rotated key for signing votes.

Reviewer 2: We should also note that the KMS/HSM should act accordingly (with the correct key) at a specific H, where H is the height after the tx is committed. Maybe this will be tricky?

Author: do you mean that the state machine combined with KMS/HSM should support auto-consensus-key-switch feature? I think it is not only tricky, but also too risky to be automated. I think key rotation deserves some manual operations considering related double-signing risk.

Reviewer 2: No no, I mean operators should take note that their KMS/HSM act and operate accordingly. Meaning, when you submit `MsgRotateConsPubKey` and it gets confirmed at height `H`, that their setup should start using the new key at height `H+1` and not the old key.

Author: Okay. I will make it clearer.

Author: done!

# ADR 016: Validator consensus key rotation

## Changelog

- 23-10-2019: initial draft

## Context

Validator consensus key rotation feature has been discussed and requested for a long time, for the sake of safer validator 
key management policy.(https://github.com/tendermint/tendermint/issues/1136) So, we suggest one of the simplest form of
validator consensus key rotation implementation mostly onto Cosmos-SDK.

## Decision

### Pseudo procedure for consensus key rotation

- create new random consensus key.
- create and broadcast a transaction(RotateValConsensusKey) that the new consensus key is now coupled with the validator operator with signature from validator wallet key.
- old consensus key becomes unable to participate on consensus immediately after the update of key mapping state on-chain.
- start validating with new consensus key.

> 

Author: Yes I agree. I will add a comment related to HSM and KMS.

- validator can store the consensus key in HSM(Hardware Security Module) for better key security policy
- or, validator also can store the consensus key in KMS(Key Management Service) and connects multiple validator with KMS for validator co-location

### Considerations

- consensus key mapping information management strategy
    - stores history of each key mapping changes in kvstore.
    - gaiad can search corresponding consensus key paired with given validator operator for any arbitrary height in recent unbonding period.
    - gaiad does not need any historical mapping information which is past more than unbonding period.
- limits
    - a validator cannot rotate its consensus key more than N time for any unbonding period, to prevent spams.
    - a validator should contribute X atoms to community fund to rotate its consensus key, also to prevent spams.
    - parameters can be decided by governance and stored in genesis file.
- slash module
    - slash module can search corresponding consensus key for any height so that it can decide which consensus key is supposed to be used for given height.
- data pruning
    - database does not need to keep the historical key mapping changes after unbonding period past.
    - when pruning is on, a node prunes all historical key mapping changes past unbonding period.

> - further implementation

Reviewer 2: I would leave this out of the ADR and initial implementation for now.

Author: OK. I will remove data pruning and further implementation part.

    - introducing "validator control key" which has authority to sign transaction for given validator.
    - validator control key can be replaced to another key by transaction(RotateValControlKey).

### Special note on implementation

> 

Reviewer 2: This section is highly lacking details IMHO. The actual process (doesn't need any pseudo-code) needs to be explicit here. Outline the entire (can be relatively high-level) process of how a rotation would occur.

e.g.

```
1. The validator generates a new consensus keypair.
2. The validator generates and signs a `MsgRotateConsensusKey` tx with their operator key.
3. `ApplyAndReturnValidatorSetUpdates` ...
```

Reviewer 7: Thanks for comments, How about this?

1. The validator generates a new consensus keypair.
2. The validator generates and signs a `MsgRotateConsPubKey` tx with their operator key and new ConsPubKey

        type MsgRotateConsPubKey struct {
        	ValidatorAddress  sdk.ValAddress  `json:"validator_address" yaml:"validator_address"`
        	NewPubKey         crypto.PubKey   `json:"new_pubkey" yaml:"new_pubkey"`
        }

3. `handleMsgRotateConsPubKey` gets `MsgRotateConsPubKey`, calls `RotateConsPubKey` with emits event
4. `RotateConsPubKey` 
    - checks if `NewPubKey` is not duplicated on `ValidatorsByConsAddr`
    - overwrites `NewPubKey` in `validator.ConsPubKey`
    - deletes old `ValidatorByConsAddr`
    - `SetValidatorByConsAddr` for `NewPubKey`
    - Add `ConsPubKeyRotationHistory` for tracking rotation

            type ConsPubKeyRotationHistory struct {
            	OperatorAddress         sdk.ValAddress `json:"operator_address" yaml:"operator_address"`
            	OldConsPubKey           crypto.PubKey  `json:"old_consensus_pubkey" yaml:"old_consensus_pubkey"`
            	NewConsPubKey           crypto.PubKey  `json:"new_consensus_pubkey" yaml:"new_consensus_pubkey"`
            	RotatedHeight           int64  `json:"rotated_height" yaml:"rotated_height"`
            }

5. `ApplyAndReturnValidatorSetUpdates` checks if there is `ConsPubKeyRotationHistory` with `ConsPubKeyRotationHistory.RotatedHeight == ctx.BlockHeight()` and if so, generates 2 `ValidatorUpdate` , one for a remove validator and one for create new validator 

        abci.ValidatorUpdate{
            PubKey: tmtypes.TM2PB.PubKey(OldConsPubKey),
            Power:  0,
        }
        
        abci.ValidatorUpdate{
            PubKey: tmtypes.TM2PB.PubKey(NewConsPubKey),
            Power:  v.ConsensusPower(),
        }

6. at `previousVotes` Iteration logic of `AllocateTokens`,  `previousVote` using `OldConsPubKey` match up with `ConsPubKeyRotationHistory`, and replace validator for token allocation
7. Migrate `ValidatorSigningInfo` and `ValidatorMissedBlockBitArray` from `OldConsPubKey` to `NewConsPubKey`

Reviewer 2: Looks reasonable. Please make the changes 👍
